pkg/services/ngalert/api/api_provisioning.go

```diff
@@ -227,7 +222,7 @@ func (srv *ProvisioningSrv) RoutePutTemplate(c *contextmodel.ReqContext, body de
 		Provenance:      determineProvenance(c),
 		ResourceVersion: body.ResourceVersion,
 	}
-	modified, err := srv.templates.SetTemplate(c.Req.Context(), c.SignedInUser.GetOrgID(), tmpl)
+	modified, err := srv.templates.UpsertTemplate(c.Req.Context(), c.SignedInUser.GetOrgID(), tmpl)
```

Comment:
What's the benefit of maintaining a separate `Upsert` method as opposed to having the "create or update" logic being in the PUT handler here?

Reply:
Backward compatibility. Current provisioning APIs use upsert logic. However, the new API will use regular CRUD. Hopefully, we can get rid of the upsert method in the future.